### PR TITLE
Pushdown `RowFilter` in `ParquetExec`

### LINKED ISSUE #3360: Support `RowFilter` in `ParquetExec`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

**Describe the solution you'd like**

arrow-rs has recently added the ability to do row-level filtering while decoding parquet files. This can dramatically reduce decoding and IO overhead when appropriately selective pruning predicates are pushed down to the table scan. We should support this in `ParquetExec`

**Describe alternatives you've considered**

If a `ParquetExec` has a `PruningPredicate` it should be "compiled" to a vector of `ArrowPredicate`(s) and supplied as a `RowFilter` to the `ParquetRecordBatchStream`. We can implement this a couple of different ways:

1. Take the pruning `Expr`, create a `PhysicalExpr` and then implement a single `ArrowPredictaeFn` which evaluates it. 
2. Break apart conjunctions in the pruning `Expr` and compile each to a separate `ArrowPredicateFn` that will be applied sequentially. We can either take the ordering as given or apply some heuristic to determine the ordering. 

Some considerations:

1. We need to be careful to avoid redundant computations. As filter predicates are pushed down in the optimizer you may end up with a predicate on the scan involving the output a scalar function and also a projection in a later stage involving the same computation. If you naively execute all filter predicates during the scan you may end up doing the same work twice. 
2. The filtering is not free. At minimum you may end up decoding the same data multiple times. 
3. Currently, if users have a custom `TableProvider` they can control what gets pushed down. Is that enough configurability? 
4. Breaking the predicate into multiple filters can help in some situations if you can get the ordering correct (cheap filter first, expensive filters after) but if you get it wrong then it can be bad. This may be amplified as we add the ability to parallelize the column decoding (i.e. as the morsel-driven scheduler progresses). 

**Additional context**

### PR DESCRIPTION
**Pull request by thinkharderdev**
# Which issue does this PR close?

Closes #3360 

 # Rationale for this change

# What changes are included in this PR?

Putting this out there for comment since it's at a point where I am more or less happy with the design and all the existing tests pass. 

The key points for this PR:

1. Introduce a helper to build a `RowFilter` from a filter `Expr` pushed down to the `ParquetExec`. 
2. If a pruning predicate is available on the `ParquetExec`, build a `RowFilter` (if we ca) and construct our `ParquetRecordBatchStream` with it. 

To build a `RowFilter` we need to go through a few steps:
1. Take the pruning `Expr` (which has been combined into a single predicate) and tear it back apart into separate predicates which are ANDed together. 
2. For each predicate, first determine whether we can use it as a row filter. We consider it valid for that purpose if it does not reference any non-primitive columns (per @tustvold suggestion) and if it does not reference any projected columns (which are not yet available). 
3. Rewrite the `Expr` to replace any columns not present in the file schema to be null literal (to handle merged schemas without involving the `SchemaAdapter`), 
4. Gather some stats to estimate the cost of evaluating the expression (currently just total size of all columns and whether all columns are sorted). 
5. Now each predicate is a `FilterCandidate` and we can sort the candidates by evaluation cost so we can apply the "cheap" filters first. 
6. Convert each candidate to an `DatafusionArrowPredicate` and build a `RowFilter` from the whole lot of them. 

TODOs:
1. Need some more specific unit tests
2. Benchmarks!
3. I can't actually figure out how to tell whether columns are sorted from the `ParquetMetadata` :)
4. Current cost estimation (purely based on compressed size of columns) is the simplest thing that could possibly work but not sure if there is a better way to do it... 

A separate conceptual question is around optimizing the number of distinct filters. In this design we simply assume that we want to break the filter into as many distinct predicates as we can but I'm not sure that is always the case given that this forces serial evaluation of the filters. I can imagine many cases where it would be better to group predicates together for evaluation. I didn't want to make the initial implementation too complicated so I punted on that for now, but eventually may want to do cost estimation at a higher level to determine the optimal grouping. 

# Are there any user-facing changes?

### CODE REVIEW

**Comment by thinkharderdev** [comment]
@Ted-Jiang @tustvold @alamb FYI. Would love any feedback you guys have on this one :)

**Comment by tustvold** [comment]
I plan to review this, time permitting, tomorrow. Very exciting 😁

**Comment by tustvold** [comment]
I plan to review this, time permitting tomorrow. Very exciting 😁

**Comment by Ted-Jiang** [comment]
@thinkharderdev cool man !👍 i will review this today !

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+/// For a given set of `Column`s required for predicate `Expr` determine whether all
+/// columns are sorted. Sorted columns may be queried more efficiently in the presence of
+/// a PageIndex.
+fn columns_sorted(
+    _columns: &HashSet<Column>,
+    _schema: &Schema,
+    _metadata: &ParquetMetaData,
+) -> Result<bool> {
+    // TODO How do we know this?
```
Comment (Ted-Jiang) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:224]: If the file with `page-index`, it will stored at `ParquetMetaData .page_indexes.boundary_order`.
I think if without page-index there is no way to find out🤔

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            projected_columns: false,
+        }
+    }
+
+    pub fn build(
+        mut self,
+        metadata: &ParquetMetaData,
+    ) -> Result<Option<FilterCandidate>> {
+        let expr = self.expr.clone();
+        let expr = expr.rewrite(&mut self)?;
```
Comment (Ted-Jiang) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:145]: why should call this func 🤔

**Comment by Ted-Jiang** [comment]
>A separate conceptual question is around optimizing the number of distinct filters. In this design we simply assume that we want to break the filter into as many distinct predicates as we can but I'm not sure that is always the case given that this forces serial evaluation of the filters. I can imagine many cases where it would be better to group predicates together for evaluation. I didn't want to make the initial implementation too complicated so I punted on that for now, but eventually may want to do cost estimation at a higher level to determine the optimal grouping.

@thinkharderdev  Agree! I remember each distinct filters will apply to the projected col with `selection`.

One thing i want to mention , when applying filter pushdowm to parquet, some `filters exprs` are `partial_filters`, it will also exits in `filer operator`. I think before all filters base on min_max are `partial_filters`(is there any situation pushDowan to parquet use `full_filters`🤔 ).

After use this row_filter i think it is a `full_filters` （we need some code change in push down rule implemention）and then we could eliminate the  `filters exprs`  in `filter operator`.🤔

File: datafusion/core/src/physical_plan/file_format/row_filter.rs (hunk 2 above)
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:145]: So this will rewrite Column expressions for columns not present in the current parquet file to `Expr::Literal(ScalarValue::Null)`. Basically we want to build the `PhysicalExpr` to only deal with the projected columns required for the filter (rahter than the entire file) so we need to basically simulate what the `SchamaAdapter` would do.

**Comment by thinkharderdev** [comment]
> > A separate conceptual question is around optimizing the number of distinct filters. In this design we simply assume that we want to break the filter into as many distinct predicates as we can but I'm not sure that is always the case given that this forces serial evaluation of the filters. I can imagine many cases where it would be better to group predicates together for evaluation. I didn't want to make the initial implementation too complicated so I punted on that for now, but eventually may want to do cost estimation at a higher level to determine the optimal grouping.
> 
> @thinkharderdev Agree! I remember each distinct filters will apply to the projected col with `selection`.
> 
> One thing i want to mention , when applying filter pushdowm to parquet, some `filters exprs` are `partial_filters`, it will also exits in `filer operator`. I think before all filters base on min_max are `partial_filters`(is there any situation pushDowan to parquet use `full_filters`🤔 ).
> 
> After use this row_filter i think it could be a `full_filters` （we need some code change in push down rule implemention）and then we could eliminate the `filters exprs` in `filter operator`.🤔 @alamb I think you are familiar with this（rewrite the push down expr）

Yes! This is I think the next phase. Once we can push down exact filters to the scan we can represent that in the `ListingTable`. The pushdown doesn't actually rewrite the filters. The existing filter `Expr` just get pushed down and it's actually `PruningPredicate` which rewrites them as min/max filters on the statistics. But they all (currently) get pushed down as inexact which means they would get executed twice (once in the scan and once again in the filter operator). If the optimizer can push down ALL the filters as exact then we can eliminate the `Filter` operator entirely (which also unlocks the possibility of pushing the limit down to the scan as well if there is one)

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 3]
```diff
@@ -358,15 +360,40 @@ impl FileOpener for ParquetOpener {
...
             let groups = builder.metadata().row_groups();
             let row_groups =
                 prune_row_groups(groups, file_range, pruning_predicate, &metrics);
 
-            let stream = builder
-                .with_projection(mask)
-                .with_batch_size(batch_size)
-                .with_row_groups(row_groups)
-                .build()?;
+            let stream = match row_filter {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:384]: It might be nicer to write this as

```
if let Some(filter) = row_filter {
    builder = builder.with_row_filter(filter)
}
```
Or something

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+
+    let mut filters: Vec<Box<dyn ArrowPredicate>> = vec![];
+
+    for candidate in indexed_candidates {
+        let filter = DatafusionArrowPredicate::try_new(candidate, file_schema, metadata)?;
+
+        filters.push(Box::new(filter));
+    }
+
+    other_candidates.sort_by_key(|c| c.required_bytes);
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:266]: Perhaps we could move this ahead of the partition, so that it also used to sort the indexed expressions?

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                expr,
+                required_bytes,
+                can_use_index,
+                projection: self.required_column_indices,
+            }))
+        }
+    }
+}
+
+impl<'a> ExprRewriter for FilterCandidateBuilder<'a> {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:165]: This is a nice use of the existing plumbing for walking expressions :+1:

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        metadata: &ParquetMetaData,
+    ) -> Result<Option<FilterCandidate>> {
+        let expr = self.expr.clone();
+        let expr = expr.rewrite(&mut self)?;
+
+        if self.non_primitive_columns || self.projected_columns {
+            Ok(None)
+        } else {
+            let required_bytes =
+                size_of_columns(&self.required_columns, self.file_schema, metadata)?;
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:151]: Why not pass required_column_indices here instead?

File: datafusion/core/src/physical_plan/file_format/row_filter.rs (hunk 1 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:224]: Parquet doesn't have a way to express multi-column sort orders that I am aware of. I thought DataFusion had some plumbing for this, but I think it doesn't extend into the physical scan operators.

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+    }
+
+    fn evaluate(&mut self, batch: RecordBatch) -> ArrowResult<BooleanArray> {
+        match self
+            .physical_expr
+            .evaluate(&batch)
+            .map(|v| v.into_array(batch.num_rows()))
+        {
+            Ok(array) => {
+                if let Some(mask) = array.as_any().downcast_ref::<BooleanArray>() {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:79]: I think the logic in parquet already handles this, and so we can skip doing this here

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+    file_schema: &Schema,
+    table_schema: &Schema,
+    metadata: &ParquetMetaData,
+) -> Result<Option<RowFilter>> {
+    let predicates = disjoin_filters(expr);
+
+    let candidates: Vec<FilterCandidate> = predicates
+        .into_iter()
+        .flat_map(|expr| {
+            if let Ok(candidate) =
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:240]: This could be written as `expr.ok().then(|| ...)` up to you

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+/// 2. Calculate data required to estimate the cost of evaluating the filter
+/// 3. Rewrite column expressions in the predicate which reference columns not in the particular file schema.
+///    This is relevant in the case where we have determined the table schema by merging all individual file schemas
+///    and any given file may or may not contain all columns in the merged schema. If a particular column is not present
+///    we replace the column expression with a literal expression that produces a null value.
+struct FilterCandidateBuilder<'a> {
+    expr: Expr,
+    file_schema: &'a Schema,
+    table_schema: &'a Schema,
+    required_columns: HashSet<Column>,
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:121]: I don't think this is necessary, in addition to required_column_indices

**Review by tustvold — COMMENTED** [review]
This is looking really cool, sorry it has taken so long to review.

Perhaps we could have an option or something to not re-order predicates, and this would allow the `TableProvider` and/or physical optimizer to use additional information to rewrite the predicates in a specific order? To give a concrete example of where this might be useful, in IOx the parquet files are sorted lexicographically, and it therefore makes sense for predicates to be evaluated in this order.

File: datafusion/core/src/physical_plan/file_format/row_filter.rs (hunk 9 above)
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:121]: yeah, I had this one first and then needed to add the indices for something else but we don't need both.

**Review by alamb — COMMENTED** [review]
TLDR I really like this PR @thinkharderdev  -- the structure is very cool. 

We probably need to think about how to test it a bit more before turning it on by default.

Perhaps we could add a flag to the parquet reader exec to enable / disable the entire "push predicates down to the parquet scan" optimization, disabled by default (rather than just the  `reorder_predicates` flag) and merge the PR in.

Then we can test  in various places (like I could test with the IOx tests) and when we are confident it is ready to be enabled, we can turn it on by default.


(as a total aside, I love the name of this branch in coralogix https://github.com/coralogix/arrow-datafusion/tree/ludicrous-speed)

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+                candidate
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    if candidates.is_empty() {
+        Ok(None)
+    } else if reorder_predicates {
```
[the remark this answers is not in the document]
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:233]: Yeah, everything else I could come up with I think ended up being pretty well-approximated by the column size. What will be interesting to figure out from an API design perspective is how to allow users to specify hints. There may be invariants or regularities in the dataset that aren't really expressed in any metadata that could still be relevant to ordering.

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:262]: I think the API in this PR is pretty good in terms of Hints: 

* Option 1 (default): Allow implementation to reorder based on heuristics
* Option 2: Do not reorder predicates and apply them in the order specified

One limitation of Option 2 is that the order can't be specified for different parquet files but if that is an important usecase we could design something more sophisticated later.

**Comment by alamb** [comment]
@thinkharderdev  please let me know when you think this PR is ready for final review / merge (it still says "RFC")

**Comment by thinkharderdev** [comment]
> @thinkharderdev please let me know when you think this PR is ready for final review / merge (it still says "RFC" in the title so I am not quite sure)

I think it's good now (pending any other feedback/comments). I can work on the benchmarks in a follow-up PR.

**Comment by alamb** [comment]
Will review carefully tomorrow

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 11]
```diff
@@ -67,6 +68,30 @@ use parquet::file::{
 };
 use parquet::schema::types::ColumnDescriptor;
 
+#[derive(Debug, Clone, Default)]
+/// Specify options for the parquet scan
+pub struct ParquetScanOptions {
+    /// If true, any available `pruning_predicate` will be converted to a `RowFilter`
+    /// and pushed down to the `ParquetRecordBatchStream`. This will enable row level
+    /// filter at the decoder level
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:76]: 👍

Comment (alamb) [reply datafusion/core/src/physical_plan/file_format/parquet.rs:76]:
```suggestion
    /// filter at the decoder level. Defaults to false
```

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 12]
```diff
@@ -839,6 +893,7 @@ mod tests {
         projection: Option<Vec<usize>>,
         schema: Option<SchemaRef>,
         predicate: Option<Expr>,
+        pushdown_predicate: bool,
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:896]: Maybe in a follow on PR we can adjust the tests to do the same scan both with and without pushdown

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 13]
```diff
@@ -1139,14 +1246,53 @@ mod tests {
...
+
+        let filter = col("c2").eq(lit(1_i64));
+
+        // read/write them files:
+        let read =
+            round_trip_to_parquet(vec![batch1, batch2], None, None, Some(filter), true)
+                .await
+                .unwrap();
+
+        // This does not look correct since the "c2" values in the result do not in fact match the predicate `c2 == 0`
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:1281]: This comment seems out of place / out of date as the predicate is `c2 == 1` and the actual rows returned are in fact where `c2 = 1`

File: datafusion/core/src/physical_plan/file_format/row_filter.rs (hunk 7 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:106]: @thinkharderdev  do you plan on making this change in this PR or a follow on on?

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+        candidates.sort_by_key(|c| c.required_bytes);
+
+        let (indexed_candidates, other_candidates): (Vec<_>, Vec<_>) =
+            candidates.into_iter().partition(|c| c.can_use_index);
+
+        let mut filters: Vec<Box<dyn ArrowPredicate>> = vec![];
+
+        for candidate in indexed_candidates {
+            let filter =
+                DatafusionArrowPredicate::try_new(candidate, file_schema, metadata)?;
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:272]: Can this ever return an error due to some sort of unsupported predicate ? If so, perhaps we could log the error (in `debug!`) and continue.

File: datafusion/expr/src/expr_fn.rs [hunk 15]
```diff
@@ -429,6 +429,40 @@ pub fn combine_filters(filters: &[Expr]) -> Option<Expr> {
     Some(combined_filter)
 }
 
+/// Take combined filter (multiple boolean expressions ANDed together)
```
Comment (alamb) [inline datafusion/expr/src/expr_fn.rs:432]: ❤️

File: datafusion/expr/src/expr_fn.rs [hunk 16]
```diff
@@ -429,6 +429,40 @@ pub fn combine_filters(filters: &[Expr]) -> Option<Expr> {
...
+/// consisting of the input filter expressions joined with logical OR.
+/// Returns None if the filters array is empty.
+pub fn combine_filters_disjunctive(filters: &[Expr]) -> Option<Expr> {
+    if filters.is_empty() {
+        return None;
+    }
+    let combined_filter = filters
+        .iter()
+        .skip(1)
+        .fold(filters[0].clone(), |acc, filter| or(acc, filter.clone()));
```
Comment (alamb) [inline datafusion/expr/src/expr_fn.rs:462]: I have also seen this pattern written with `reduce` -- so something like (untested)

```suggestion
    let combined_filter = filters
        .iter()
        .reduce(|acc, filter| or(acc, filter.clone()));
```

**Review by alamb — APPROVED** [review]
I think this PR can be merged in as is and we can keep iterating on it subsequently. 

Really nice @thinkharderdev  and kudos to the rest of the predicate pushdown team @Ted-Jiang  and @tustvold  (sorry if I have forgotten others)

It would be nice to file a follow on issue (I can do so if you like) listing the steps we felt were necessary to turn on predicate pushdown by default. In my mind this is primarily about testing (functional and performance).

**Review by alamb — COMMENTED** [review]
Since this feature defaults to off, I think it would be ok to merge in prior to the 12.0.0 release candidate but I defer to @andygrove  on that one

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+
+            filters.push(Box::new(filter));
+        }
+
+        Ok(Some(RowFilter::new(filters)))
+    }
+}
+
+fn is_primitive_field(field: &Field) -> bool {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:298]: Filed https://github.com/apache/arrow-rs/issues/2704

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:298]:
```suggestion
/// return true if this is a non nested type.
// TODO remove after https://github.com/apache/arrow-rs/issues/2704 is done
fn is_primitive_field(field: &Field) -> bool {
```

**Comment by thinkharderdev** [comment]
> I think this PR can be merged in as is and we can keep iterating on it subsequently.
> 
> Really nice @thinkharderdev and kudos to the rest of the predicate pushdown team @Ted-Jiang and @tustvold (sorry if I have forgotten others)
> 
> It would be nice to file a follow on issue (I can do so if you like) listing the steps we felt were necessary to turn on predicate pushdown by default. In my mind this is primarily about testing (functional and performance).

Thanks! I can clean up the last few things mentioned in your review this afternoon/evening. I'll also create a followup ticket with remaining items. In general I think the follow-ons are:

1. Add some proper benchmarks 
2. Once we are on arrow-rs 23 (I think) and using the offset index is stabilized we should allow users to pass that down in the `ParquetScanOption`.
3. Once we are all comfortable with it, turn filter pushdown on by default.

File: datafusion/core/src/physical_plan/file_format/row_filter.rs (hunk 14 above)
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:272]: It shouldn't ever fail unless there is a bug (i.e. there is no expected failure case). I think if this fails then the query will almost certainly fail somewhere else.

File: datafusion/core/src/physical_plan/file_format/parquet.rs (hunk 13 above)
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/file_format/parquet.rs:1281]: Yeah, lazy copy pasta :)

File: datafusion/core/src/physical_plan/file_format/row_filter.rs (hunk 7 above)
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:106]: I think @tustvold was referring to the `prep_null_mask` call which is removed. Unless I'm missing something then this should be good to go.

**Comment by thinkharderdev** [comment]
Created issues for follow-up work:

https://github.com/apache/arrow-datafusion/issues/3456
https://github.com/apache/arrow-datafusion/issues/3457

**Comment by thinkharderdev** [comment]
I'm not sure what's going on with the checks. The errors look like it is building a different version of the code....

**Comment by tustvold** [comment]
Github actions merges master when running tests, try merging master into this branch perhaps?

**Comment by alamb** [comment]
I collected follow on work and other related items I could find into a tracking ticket / epic in https://github.com/apache/arrow-datafusion/issues/3462 -- it is great story / project.

**Comment by alamb** [comment]
🚀

File: datafusion/expr/src/expr_fn.rs [hunk 19]
```diff
@@ -698,4 +730,56 @@ mod test {
...
+            Field::new("b", DataType::Utf8, true),
+            Field::new("c", DataType::Utf8, true),
+        ]);
+
+        let expr = and(col("a"), col("b"));
+        let actual = uncombine_filter(expr);
+
+        assert_predicates(actual, vec![col("a"), col("b")]);
+
+        let expr = col("a").and(col("b")).or(col("c"));
```
Comment (liukun4515) [inline datafusion/expr/src/expr_fn.rs:780]: we should add other test case for this.
The expr is ` a and b or c` which will be handle lik `binary( a and b, or c).
But I think the above expr can be converted to `c or a and b`, and will get error result.

cc @thinkharderdev @alamb

File: datafusion/core/src/physical_plan/file_format/row_filter.rs [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+            } else if self.table_schema.index_of(&column.name).is_err() {
+                // If the column does not exist in the (un-projected) table schema then
+                // it must be a projected column.
+                self.projected_columns = true;
+            }
+        }
+        Ok(RewriteRecursion::Continue)
+    }
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
```
Comment (liukun4515) [inline datafusion/core/src/physical_plan/file_format/row_filter.rs:202]: @alamb @thinkharderdev 
I find a issue caused by this rewrite in my work https://github.com/apache/arrow-datafusion/pull/3396
if we have two parquet, one has c1,c2, the other has c1,c3.
If the filter is c2 = 1, then it will produce a expr  NULL = 1 for the parquet with c1,c3 column.
After I remove the binary type coercion in the physical phase, the binary physical expr of `NULL = INT(1)` can't be created, because we don't need to support type coercion in creation physical expr.

Comment (liukun4515) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:202]: We need to make sure the data type of the NULL value.

Comment (liukun4515) [reply datafusion/core/src/physical_plan/file_format/row_filter.rs:202]: we need to get the data type from the table schema.

File: datafusion/core/src/physical_plan/file_format/parquet.rs [hunk 21]
```diff
@@ -358,6 +397,21 @@ impl FileOpener for ParquetOpener {
                 adapted_projections.iter().cloned(),
             );
 
+            if let Some(predicate) = pushdown_filters
+                .then(|| pruning_predicate.as_ref().map(|p| p.logical_expr()))
+                .flatten()
+            {
+                if let Ok(Some(filter)) = build_row_filter(
```
Comment (liukun4515) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:404]: if there is a error when build the row filter, why not throw the error out?

File: datafusion/expr/src/expr_fn.rs (hunk 19 above)
Comment (liukun4515) [reply datafusion/expr/src/expr_fn.rs:780]: > Could you be a more specific @liukun4515 -- I am not sure if you are pointing out a potential bug . This line makes the predicate `(a AND b) OR c` I believe. What do you mean by `c or a and b`?

sorry for that, I think  it's my wrong for this.
